domain-models-runtime/src/main/resources/templates/db_scripts/audit.ftl

```diff
@@ -1,64 +1,61 @@
-
--- audit table to keep a history of the changes
--- made to a record.
-CREATE TABLE IF NOT EXISTS ${myuniversity}_${mymodule}.audit_${table.tableName} (
-   id UUID PRIMARY KEY,
-   orig_id UUID NOT NULL,
-   operation char(1) NOT NULL,
-   jsonb jsonb,
-   created_date timestamp not null
-   );
-
+-- trigger for the audit table to keep a history of the changes made to a record.
 CREATE OR REPLACE FUNCTION ${myuniversity}_${mymodule}.audit_${table.tableName}_changes() RETURNS TRIGGER AS $${table.tableName}_audit$
-    DECLARE
-    <#if table.auditingSnippet??>
-      <#if table.auditingSnippet.delete??>
-      ${table.auditingSnippet.delete.declare}
+  DECLARE
+  <#if table.auditingSnippet??>
+    <#if table.auditingSnippet.delete??>
+    ${table.auditingSnippet.delete.declare}
+    </#if>
+    <#if table.auditingSnippet.update??>
+    ${table.auditingSnippet.update.declare}
+    </#if>
+    <#if table.auditingSnippet.insert??>
+    ${table.auditingSnippet.insert.declare}
+    </#if>
+  </#if>
+    seed TEXT;
+    maxid UUID;
+    jsonb JSONB;
+  BEGIN
+    maxid = (SELECT ${myuniversity}_${mymodule}.max(id) FROM ${myuniversity}_${mymodule}.audit_${table.tableName});
+    IF maxid IS NULL THEN
+      -- don't use random, that will be different on each pg-pool replication
+      seed = md5(concat('${myuniversity}_${mymodule}.audit_${table.tableName}', NEW.jsonb));
+      -- avoid overflow
+      IF (left(seed, 13) = 'ffffffff-ffff') THEN
+        seed = overlay(seed placing '00000000-0000' from 1);
+      END IF;
+      -- UUID version byte
+      seed = overlay(seed placing '4' from 13);
+      -- UUID variant byte
+      seed = overlay(seed placing '8' from 17);
+      maxid = seed::uuid;
+    ELSE
+      maxid = ${myuniversity}_${mymodule}.next_uuid(maxid);
+    END IF;
+    jsonb = CASE WHEN TG_OP = 'DELETE' THEN OLD.jsonb ELSE NEW.jsonb END;
+    jsonb = jsonb_set(jsonb, '{origId}', jsonb->'id');
+    jsonb = jsonb_set(jsonb, '{id}', to_jsonb(maxid::text));
+    jsonb = jsonb_set(jsonb, '{operation}', to_jsonb(left(TG_OP, 1)));
+    jsonb = jsonb_set(jsonb, '{createdDate}', to_jsonb(current_timestamp::text));
```

Comment:
There is a chance the original json already contains these field name. Would it be safer to make these field name a bit more unique like `audit_orig_id`, 'audit_operation`, and `audit_creation_date`?